### PR TITLE
47022173: 3D Segmenting prostate MRI scans using 3D Improved UNet

### PR DESCRIPTION
**Pull request by 27abdullah**
Task 4 - Hard Difficulty. This project uses a deep learning model, the 3D improved UNet, to segment 3D MRI scans of the prostate. The model segments the input MRI scans into background, body, bone, bladder, rectum, prostate. This project aimed to achieve a dice score of greater than 0.7 across all labels.  

Added Files:
- dataset.py - Contains the ProstateDataset3D class to load and process the images and masks for a dataloader.
- driver.py - The driver code to run training or inference using the command line.
- modules.py - Contains the Modified3DUNet class which is the model for this task. It also has a function to init weights for the model.
- predict.py - Contains example usage of the code and visualisations.
- train.py - Contains the validate and train loops (as functions) and a function to save predictions for checkpoints.
- utils.py - Contains various constants used throughout the code e.g., default hyperparameters, model dimensions.
- requirements.txt - Dependencies.
- assets - Folder containing images for readme.

### CODE REVIEW

**Comment by celinsze** [comment]
**This is an initial inspection, no action is required at this point**
Recognition problem:
Problem Solving: (5)
•	Algorithm solves the problem and achieves a Dice score of above 0.7 for all labels. 
Algorithm implementation: (3)
•	Algorithm implemented in PyTorch for the improved UNet3D model and functions.
Design: Good coding design (1)
Commenting: Clear and sufficient comments throughout the code (1)
Difficulty: 3D Improved UNet – Hard (10)

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard. UNet3D (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time submission.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>0</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by shakes76** [comment]
Segmentations not one hot encoded -2 additionally

**Comment by shakes76** [comment]
Good work
